### PR TITLE
Add new 1.9 block entities

### PR DESCRIPTION
**Pull request by kashike**
[**API**](https://github.com/SpongePowered/SpongeAPI/pull/1216) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/691)
Supersedes #1146

### CODE REVIEW

**Comment by Aaron1011** [comment]
@kashike: Is this ready to merge?

File: src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/EndGatewayData.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ */
+public interface EndGatewayData extends DataManipulator<EndGatewayData, ImmutableEndGatewayData> {
+
+    /**
+     * Gets the {@link Value} for the exit portal location of
+     * the {@link EndGateway}.
+     *
+     * @return The value for the exit portal location
+     */
+    Value<Vector3i> exitPortal();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/EndGatewayData.java]: exitLocation

Comment (gabizou) [reply src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/EndGatewayData.java]: Better yet, position

Comment (Zidane) [reply src/main/java/org/spongepowered/api/data/manipulator/mutable/tileentity/EndGatewayData.java]: Agreed, exitPosition

**Comment by kashike** [comment]
@Zidane
- [x] `EndGatewayData`
  - [x] rename `exitPortal` to `exitPosition`
  - [x] add `age` (`long`)
- [x] `StructureData`
  - [x] add `mode` (`Mode` - catalog type: `SAVE`, `LOAD`, `CORNER`, `DATA`)
  - [x] add `author` (`String`)
  - [x] add `ignoreEntities` (`boolean`)
  - [x] add `powered` (`boolean`)
  - [x] add `showAir` (`boolean`)
  - [x] add `showBoundingBox` (`boolean`)
  - [x] add `integrity` (`float`)
  - [x] add `seed` (`long`)
  - [x] add `rotation` (see `net.minecraft.util.Rotation`)
  - [x] add `mirror` (see `net.minecraft.util.Mirror`)
  - [x] add `metadata` (`String`)

File: src/main/java/org/spongepowered/api/block/tileentity/EndGateway.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import org.spongepowered.api.data.manipulator.mutable.tileentity.EndGatewayData;
+
+/**
+ * Represents an End Gateway.
+ *
+ * @see EndGatewayData
+ */
+public interface EndGateway extends TileEntity {
+
+    default EndGatewayData getEndGatewayData() {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/block/tileentity/EndGateway.java:42]: Needs some javadocs.

File: src/main/java/org/spongepowered/api/block/tileentity/Structure.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+package org.spongepowered.api.block.tileentity;
+
+import org.spongepowered.api.data.manipulator.mutable.tileentity.StructureData;
+
+/**
+ * Represents a Structure.
+ */
+public interface Structure extends TileEntity {
+
+    default StructureData getStructureData() {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/block/tileentity/Structure.java:44]: Likewise here.

File: src/main/java/org/spongepowered/api/block/tileentity/Structure.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.block.tileentity;
+
+import org.spongepowered.api.data.manipulator.mutable.tileentity.StructureData;
+
+/**
+ * Represents a Structure.
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/block/tileentity/Structure.java:31]: What does a `Structure` tile do? Explain that it has the capability of saving, pasting, and clearing schematics.

File: src/main/java/org/spongepowered/api/data/type/StructureMode.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.data.type;
+
+import org.spongepowered.api.CatalogType;
+import org.spongepowered.api.util.annotation.CatalogedBy;
+
+@CatalogedBy(StructureModes.class)
+public interface StructureMode extends CatalogType {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/data/type/StructureMode.java:35]: Javadocs?

File: src/main/java/org/spongepowered/api/data/type/StructureModes.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.data.type;
+
+import org.spongepowered.api.util.generator.dummy.DummyObjectProvider;
+
+public final class StructureModes {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/data/type/StructureModes.java:32]: Javadocs?

File: src/main/java/org/spongepowered/api/data/type/StructureModes.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    public static final StructureMode LOAD = dummy("LOAD");
+
+    public static final StructureMode SAVE = dummy("SAVE");
+
+    // SORTFIELDS:OFF
+
+    private StructureModes() {
+    }
+
+    private static StructureMode dummy(String fieldName) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/data/type/StructureModes.java]: No. Just use the traditional long comment of making these fake fields.

**Pull request merged**